rfcs/0008-distributed-dns.md

```diff
@@ -0,0 +1,274 @@
+
+This clusterID is used to identify which A/CNAME records were created by which cluster/controller. This is done via the registry implementation in external-dns. Initially we will use the TXT registry, but in the future we will also want to offer a different form of storage for the registry .
+
+The suggestion [here](https://groups.google.com/g/kubernetes-sig-architecture/c/mVGobfD4TpY/m/nkdbkX1iBwAJ) is to use the UID of the `kube-system` namespace, as this [cannot be deleted easily](https://groups.google.com/g/kubernetes-sig-architecture/c/mVGobfD4TpY/m/lR4SnaqpAAAJ) - so is practically speaking indelible.
+
+We can take the UID of that namespace and apply a hashing algorithm to result in 6-7 character code which can be used as the clusterID for the local cluster. This will be applied 
+
+### DNS Operator
+
+This component takes the DNSRecord output by the kuadrant-operator and ensure they are accurately reflected in the DNS Provider zone and currently assumes the DNS Record is the source of truth and always ensures that the zone in the DNS Provider matches the local DNS Record CR.
+
+There are several changes required in this component:
+
+#### Ensure local records are accurate
+
+When the local DNSRecord is updated, the DNS Operator will ensure those values are present in the zone by interacting directly with the DNS Provider. As mentioned above it will pull the zone, merge in its changes and write back the result before requeuing the resource for validation. With each requeuing it will add a random amount of time to attempt to fall out sync with any conflicting controllers.
+
+#### Cleanup
+
+When a DNS Policy is marked for deletion the kuadrant operator will delete all relevant DNSRecords.
+
+Whenever a deleted DNSRecord is reconciled by the DNS Operator, it will remove any relevant records from the DNS Provider (i.e. any record identified by the clusterID for the deleted host.
+
+#### Prune dead ends from DNS Records
+
+If a DNSRecord is deleting, there is the potential that we are leaving behind a dead end, which will need to be recursively cleaned until none remain.
+
+What is a dead end? If a CNAME exists whose value is a record that does not exist, this CNAME will not resolve.
+
+The controller will need to work through the DNS Record and ensure that from hostname to A (or external CNAME) records is structurally sound and that any dead ends are removed, to do this, the DNS Operator will need to first read what is published in the provider zone.
+
+#### Removing the finalizer from DNSRecord
+
+When a DNSRecord is reconciled, the DNS Operator should apply a finalizer to it.
+
+When a DNSRecord is being removed, the following must be successfully completed before the finalizer can be removed:
+- Remove the local clusters records and targets from the relevant zone
+- Perform a prune on the relevant root host
+- Apply the results of the prune to the DNS Provider
```

Comment:
also delete healthchecks if needed (how to detect it - no records left?)

Reply:
This is a good point. @philbrookes any thoughts on when we delete a health check? I would think it is part of the dead end check? IE if there is no endpoints for a given rootHost then remove?

Reply:
had to change the rfc number which is why this shows as outdated